design/design.md

```diff
@@ -0,0 +1,414 @@
+nodes to follow the right chain.
+A natural method for this certification is to use a
+weighted threshold multisignature scheme constructed using decentralized ALBA.
+
+Ouroboros Leios aims to increase the throughput of the blockchain in terms of
+the number of transactions per second.
+It does so by decoupling the block payloads from the consensus mechanism which
+operates only on references to blocks.
+To make sure that the chosen block's payload is correct and available
+(which potentially could be not due to malicious behavior),
+all blocks in the input to the consensus protocol need to have
+a certificate of availability proving that at least one honest node has verified
+the block content and is available to distribute it.
+A natural solution to certify data availability would be to use a
+weighted threshold multisignature scheme constructed using decentralized ALBA.
+
+Cross-chain bridges are protocols allowing one blockchain to verify
+information on another blockchain.
+In essense this is the same problem that the Mithril solves with a potentially
+added constraint that on-chain verification must be computationally cheap.
+Whereas Mithril is specific to the Cardano blockchain, one can envision
+a similar protocol, powered by ALBA, for other cryptocurrencies including
+partner chains like Midnight.
+
+Applications of ALBA are not limited to Cardano and blockchains in general.
+For instance, the [paper](https://eprint.iacr.org/2023/1655) describes how
+this primitive can be used to optimize the prover running time of
+a Universally Composable SNARK.
+
+## Requirements
+
+We intend to make an implementation of ALBA that can be used in
+Mithril, Peras, Leios, cross-chain bridges and potentially other protocols
+inside and outside the Cardano universe.
+Our implementation will mostly be guided by our main "clients":
+Mirthril, Peras and Leios.
+
+ALBA is a generic tool that lets one prove knowledge of any kind of data
+elements, but
+as noted in (todo: alba applications section), these three applications
+don't need the full expressivity of ALBA, they really need
+a multisignature scheme, where the data elements would specifically be
+signatures.
+Thus there is question of whether we want our library to provide a general
+ALBA interface or simply an interface for a multisignature scheme.
+We will do both.
+1. We will implement the most generic version of ALBA to make it suitable for
+as many applications as possible.
+This means that there is only a generic notion of "data elements" and no notion
+of signatures, VRF, etc.
+2. We will use the generic ALBA library to implement a multisignature scheme,
+either as a small example in this repository or a full-fledged library in a
+separate repository.
+Either way, the multi-signature implementation would serve as a flagship
+example of how the ALBA library can be used.
+
+The requirements for our (general) ALBA implementation can be roughly split
```

Comment:
These requirements are very, very loose. I would rather have goals written: 
- the library must implement correctly the protocols as written in the paper as a reference implementation.
- the code must be easy to follow
...

Reply:
So how do we know if the requirements are satisfied or not?